presidio-analyzer/presidio_analyzer/predefined_recognizers/au_abn_recognizer.py

```diff
@@ -0,0 +1,90 @@
+from typing import Optional, List, Tuple
+
+from presidio_analyzer import Pattern, PatternRecognizer
+
+
+class AuAbnRecognizer(PatternRecognizer):
+    """
+    Recognizes Australian Business Number ("ABN").
+
+    Recognizes ABN using regex, context words, and checksum.
+
+    :param patterns: List of patterns to be used by this recognizer
+    :param context: List of context words to increase confidence in detection
+    :param supported_language: Language this recognizer supports
+    :param supported_entity: The entity this recognizer can detect
+    :param replacement_pairs: List of tuples with potential replacement values
+    for different strings to be used during pattern matching.
+    This can allow a greater variety in input, for example by removing dashes or spaces.
+    """
+
+    PATTERNS = [
+        Pattern(
+            "ABN (Medium)",
+            r"\b\d{2}\s\d{3}\s\d{3}\s\d{3}\b",
+            0.5,
+        ),
+        Pattern(
+            "ABN (Low)",
+            r"\b\d{11}\b",
+            0.3,
+        ),
+    ]
+
+    CONTEXT = [
+        "australian business number",
+        "abn",
+    ]
+
+    def __init__(
+        self,
+        patterns: Optional[List[Pattern]] = None,
+        context: Optional[List[str]] = None,
+        supported_language: str = "en",
+        supported_entity: str = "AU_ABN",
+        replacement_pairs: Optional[List[Tuple[str, str]]] = None,
+    ):
+        self.replacement_pairs = (
+            replacement_pairs if replacement_pairs else [("-", ""), (" ", "")]
+        )
+        context = context if context else self.CONTEXT
+        patterns = patterns if patterns else self.PATTERNS
+        super().__init__(
+            supported_entity=supported_entity,
+            patterns=patterns,
+            context=context,
+            supported_language=supported_language,
+        )
+
+    def validate_result(self, pattern_text: str) -> bool:
+        """
+        Validate the pattern logic e.g., by running checksum on a detected pattern.
+
+        :param pattern_text: the text to validated.
+        Only the part in text that was detected by the regex engine
+        :return: A bool indicating whether the validation was successful.
+        """
+        # Pre-processing before validation checks
+        text = self.__sanitize_value(pattern_text, self.replacement_pairs)
+        abn_list = [int(digit) for digit in text]
+
+        # Set weights based on digit position
+        weight = [10, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19]
+
+        # Perform checksums
+        abn_list[0] = 9 if abn_list[0] == 0 else abn_list[0] - 1
```

Comment:
The checksum logic for all these new entities seem very similar (please correct me if I'm wrong). Could we perhaps create a class with default logic and have the different AU entity recognizers inherit from it?

Reply:
There are similarities as all the checksum use a sum product and modulo operation. However, the following differs:

- Check digit appears at different locations for each entity
- Weights are different across the different entities
- Modulo number is different.
- AU_ACN checks the complement of the remainder of Modulo, rather than just remainder. 

Given the above, do you recommend creating a class?

Reply:
For the first three bullets I would say yes, but I assume that the fourth changes the logic, so I assume it would be an overkill to create a class. Let's keep it as is.